proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -8329,153 +8344,82 @@ lemma valid_refills_r_amount_bounded_max_sc_period:
-  \<Longrightarrow> r_amount x \<le> sc_budget sc"
-  apply (clarsimp simp: sc_valid_refills_def rr_valid_refills_def word_le_nat_alt split: if_splits)
-  apply (rule_tac y="unat (refills_sum (sc_refills sc))" in order_trans[rotated], clarsimp)
-  apply (subst refills_sum_def)
-  apply (subst valid_refills_unat_sum_list, simp, simp)
-  apply (rule member_le_sum_list, simp)
-  done
-
-lemma sc_valid_refills_hd_in_list:
-  "sc_valid_refills sc
-  \<Longrightarrow> hd (sc_refills sc) \<in> set (sc_refills sc)"
-  by (case_tac "sc_refills sc";
-      clarsimp simp: sc_valid_refills_def rr_valid_refills_def split: if_splits)
-
-lemma sc_refill_ready_release_time_well_bounded:
-  "kheap s p = Some (SchedContext sc n)
-   \<Longrightarrow> cur_sc s = p
-   \<Longrightarrow> current_time_bounded 2 s
-   \<Longrightarrow> valid_refills p s
-   \<Longrightarrow> sc_refill_ready (cur_time s) sc
-   \<Longrightarrow> unat (r_time (refill_hd sc)) + unat MAX_PERIOD + unat MAX_PERIOD \<le> unat max_time"
-  apply (clarsimp simp: current_time_bounded_def refill_ready_def mult_2 add.assoc[symmetric])
-  apply (erule order_trans[rotated], clarsimp)
-  apply (rule order_trans[OF _ unat_plus_gt])
-  apply (simp add: word_le_nat_alt)
-  done
-
-lemma schedule_used_unat_sum:
-  "sum_list (map unat (map r_amount list)) + unat (r_amount new) \<le> unat max_time
-   \<Longrightarrow> sum_list (map unat (map r_amount (schedule_used full list new)))
-       = sum_list (map unat (map r_amount list)) + unat (r_amount new)"
-  supply map_map[simp del]
-  apply (cases list)
-   apply (clarsimp simp: schedule_used_def)
-  apply (rename_tac a lista)
-  apply (case_tac "can_merge_refill (last list) new")
-   apply (clarsimp simp: schedule_used_def)
-   apply (intro conjI impI)
-    apply (simp add: unat_plus_simple unat_sum_bound_equiv)
-   apply (subst unat_add_lem')
-    apply (prop_tac "unat (r_amount (last lista)) \<le> sum_list (map unat (map r_amount lista))")
-     apply (simp add: member_le_sum_list)
-    apply (clarsimp simp: max_word_def)
-   apply simp
-   apply (prop_tac "butlast lista @ [last lista] = lista", fastforce)
-   apply (prop_tac "sum_list (map unat (map r_amount [last lista])) = unat (r_amount (last lista))")
-    apply simp
-   apply (metis sum_list_append map_append)
-  apply (clarsimp simp: schedule_used_def)
-  apply (intro conjI impI)
-   apply (subst unat_add_lem'; fastforce?)
-   apply (clarsimp simp: max_word_def)
-  apply (subst unat_add_lem')
-   apply (prop_tac "unat (r_amount (last lista)) \<le> sum_list (map unat (map r_amount lista))")
-    apply (simp add: member_le_sum_list)
-   apply (clarsimp simp: max_word_def)
-  apply simp
-  apply (prop_tac "butlast lista @ [last lista] = lista", fastforce)
-  apply (prop_tac "sum_list (map unat (map r_amount [last lista])) = unat (r_amount (last lista))")
-    apply simp
-  apply (metis sum_list_append map_append)
+lemma unat_max_time:
+  "2 ^ LENGTH(64) - 1 = unat max_time"
```

Comment:
This and the next lemma should probably be about `max_word` instead of `max_time`, and also marked for being moved (possible even into `Word_Lib`, although it's hard to tell whether they're distinct enough from the already existing lemmas)

Reply:
Sure. I've got some nice rules generalised to `'a :: len word`. Maybe `Word_Lemmas_Internal.thy` in `Word_Lib`?

Reply:
Yes, the `*_Internal` theories are the right place to put these.

Reply:
Apart from this one, of course :-) (sorry, saw that as soon as I wrote it). `max_time` is not going to be defined there yet, so somewhere close to where `max_time` is defined would be a good spot.

Reply:
(when generalised to `max_word`)